### PR TITLE
List Feature

### PR DESCRIPTION
**Pull request by auamod**

### CODE REVIEW

File: lc-daemon/main.go [hunk 1]
```diff
@@ -194,40 +199,31 @@ func obtainStatus(c chan []string, p int, wg *sync.WaitGroup) {
...
 
 	for 0 == strings.Index(text, "CLIENT_LIST") {
 		strList := strings.Split(text, ",")
 		c <- strList
+		text, _ = reader.ReadString('\n')
 	}
-
-	// XXX maybe it is easier to just close the connection, who cares about
-	// quit?
-	fmt.Fprintf(conn, "quit\n")
```
Comment (fkooman) [inline lc-daemon/main.go]: the connection does not need to be closed?

File: lc-daemon/main.go [hunk 2]
```diff
@@ -194,40 +199,31 @@ func obtainStatus(c chan []string, p int, wg *sync.WaitGroup) {
...
 
-	// disconnect the client
+	// send status command to OpenVPN management interface
 	fmt.Fprintf(conn, "status 2\n")
 	text, _ := reader.ReadString('\n')
 	for 0 != strings.Index(text, "CLIENT_LIST") {
 		// walk until we find CLIENT_LIST
-		text, _ = reader.ReadString('\n')
+		// exit loop if no clients are found -> if not inf loop searching for "CLIENT_LIST"
+		if 0 != strings.Index(text, "END") {
```
Comment (fkooman) [inline lc-daemon/main.go]: you can also rewrite that without `else`

Comment (fkooman) [reply lc-daemon/main.go]:
```
if 0 == strings.Index(text, "END") {
    break;
}
text, _ = reader.ReadString('\n')

**Comment by auamod** [comment]
Accidentally resolved conversation
But the defer conn.Close() on line 207 already does this.
https://golang.org/doc/effective_go.html#defer

File: lc-daemon/main.go (hunk 1 above)
Comment (auamod) [reply lc-daemon/main.go]: defer conn.Close() on line 207 already does this.
https://golang.org/doc/effective_go.html#defer

Comment (fkooman) [reply lc-daemon/main.go]: Ah, but then it can also go away in the disconnect function?

Comment (auamod) [reply lc-daemon/main.go]: Yes, should be no problem. No matter what happens the connection wil be closed

File: lc-daemon/main.go (hunk 2 above)
Comment (auamod) [reply lc-daemon/main.go]: fixed in https://github.com/letsconnectvpn/lc-daemon/pull/3/commits/d45c5de0eeb5f1cb7e2b83a4d0e81258a887ca3a

**Comment by auamod** [comment]
LIST function can support upto 9999 connected clients. Before was only the amount of ports entered.
LIST also check for ports, to free up memory if no ports has been added yet

**Comment by fkooman** [comment]
> LIST function can support upto 9999 connected clients. Before was only the amount of ports entered.
> LIST also check for ports, to free up memory if no ports has been added yet

Why is this? What do you mean 9999 connected clients? Who can exactly do what?

File: lc-daemon/main.go [hunk 3]
```diff
@@ -194,40 +204,30 @@ func obtainStatus(c chan []string, p int, wg *sync.WaitGroup) {
...
-		// so no need to disconnect clients there ;-)
+		// so no need to retrieve clients there ;-)
 		c <- nil
 		return
 	}
 
 	defer conn.Close()
 
-	// turn off live OpenVPN log that can confuse our output parsing
-	fmt.Fprintf(conn, fmt.Sprint("log off\n"))
```
Comment (fkooman) [inline lc-daemon/main.go]: why did you remove this? are you sure that's safe and the output of a command can't be interleaved by log messages, i.e. when a client connects?

File: lc-daemon/main.go [hunk 4]
```diff
@@ -114,9 +114,12 @@ func handleConnection(conn net.Conn) {
...
 		if 0 == strings.Index(msg, "LIST") {
 
-			fmt.Println("LIST")
+			//prevent connection getting stuck, wait for next line
+			if len(intPortList) == 0 {
+				continue
+			}
 
-			c := make(chan []string, len(intPortList))
+			c := make(chan []string, 9999)
```
Comment (fkooman) [inline lc-daemon/main.go]: 9999 is pretty arbitarary, why does `len(intPortList)` not work?

Comment (auamod) [reply lc-daemon/main.go]: The second argument defines the size of the channel.
Even if we give it 20ports, thats quite a lot already, the channel can only accept 20 clients. If there is more than 20 clients, the channel will be blocked and a deadlock will occur.

And we can't leave it with no size, it will behave as the channel is already full.
The number should be a practical maximum of the connected clients over all the ports

Comment (fkooman) [reply lc-daemon/main.go]: but... you know exactly how many connections there will be, namely the number of ports specified in the SET_XX command... so I don't see the problem?

Comment (auamod) [reply lc-daemon/main.go]: Connections != # of ports

Here you have 3 connections to a single ports, could be more if I had more devices:
Annotation 2019-10-16 224851

And some ports can have no connections at all:
Annotation 2019-10-16 224930

These images were taken at the same time, so we have 3 clients and 2 ports. If I were to run it with just `len(intPortList)` the program will enter a deadlock, because channel could hold only 2 clients and will remain waiting.

Comment (fkooman) [reply lc-daemon/main.go]: This doesn't sound right. Why would you create a new channel for every client connection to OpenVPN? The channel is to send data back to the caller of the go routine, the caller merges all data coming through the channel(s). So one channel per OpenVPN process the daemon is talking to. Or what am I missing?

Comment (fkooman) [reply lc-daemon/main.go]: The "DISCONNECT" it doing it exactly like this, but there it is only counting successful disconnects... and that seems to work...

File: lc-daemon/main.go (hunk 3 above)
Comment (auamod) [reply lc-daemon/main.go]: With log on:
Annotation 2019-10-16 230836

With log off:
Annotation 2019-10-16 230416

As you can see the console will remain cluttered with information. We are only interested in the line with header "CLIENT_LIST" for this function. A few lines below in the code we are iterating through that information looking for that specific header.

Will still add a few lines to thoroughly go through all the information, to find all clients (in case interleaving does happen), which will stop if END is encountered.

File: lc-daemon/main.go (hunk 4 above)
Comment (auamod) [reply lc-daemon/main.go]: It's only creating 1 channel. The go-routines(depends on #of ports) will use the same channel.
Further you could say we are filling this channel with client connections(in the beginning its empty). This channel has a capacity(what we are discussing right now).

Lets say we have the default configuration; 2ports and 5client connection per port.
With `len(intPortList)` the channel will have a capacity of 2 units. But every connection is a unit sent to the channel.If the channel becomes full, the others will have to wait(in this case forever).

Comment (auamod) [reply lc-daemon/main.go]: I wanted to leave the DISCONNECT out of this PR, but here goes.
The disconnect function is working because each port cannot have duplicate common names(So unique common names in each port), this is in the default configuration. You only have to search each port once to see if the common name is there.

However if we allow multiple clients with the same CN to connect to the port: with duplicate-cn in the server files, this DISCONNECT function will no longer work.

Example with multiple same CN on 1 port:
Annotation 2019-10-17 001325

If we kill the CN in a management interface, it will return 4
Annotation 2019-10-17 001449

Thats the flaw with DISCONNECT. The daemon will only check for the 'SUCCESS' and report that only 1 has disconnected, while in truth 4 has been disconnected:
Annotation 2019-10-17 002400

So instead of only having to send 1 boolean to the channel, you are now required to send 4, thus also causing a deadlock

Comment (fkooman) [reply lc-daemon/main.go]: Ah, I think I get it now! Is there a better way to do this, instead of having to specify this `9999`? Maybe introduce a data type that contains all client info before sending it back through the channel?

File: lc-daemon/main.go (hunk 3 above)
Comment (auamod) [reply lc-daemon/main.go]: Update: https://github.com/letsconnectvpn/lc-daemon/pull/3/commits/e2b4e8444345c4963e6cb4a8603150e055204fc1

Will go through all messages searching for clients. In the case that interleaving does happen, it will continue to iterate finding all client information.

With echo/log/state on it will continue working now:
list

File: lc-daemon/main.go (hunk 4 above)
Comment (auamod) [reply lc-daemon/main.go]: Update https://github.com/letsconnectvpn/lc-daemon/pull/3/commits/0fcfd34414eb49ca227454d2f739094700c47d9f

Now each port will send one item to the channel, this item being a array of structs containing their respective connection information.
Now the arbitrary `9999` can be replaced with the previous value `len(intPortList)`
